### PR TITLE
Enhancement/query search users

### LINKED ISSUE #104: Query - Search users
## The story:
As a **social person** that **wants to find my friends on CM**, I want to **search via a name or email for people I know**

---

## What we should do:
We should **build a `searchUsers` query** so that **users can search for other users (to _eventually_ add them as a friend**.

---

## To be successful we need to:
- [ ] Have a `searchUsers` query
- [ ] Return a list of `User` entities on the result
- [ ] Should allow searching via email
- [ ] Should allow searching via name
- [ ] Contain relevant tests for work done that pass in CI

---

### Query: `searchUsers`
```gql
query ($search: String!) {
  searchUsers(search: $search) {
    id
    firstName
    lastName
    # ...
  }
}
```

### PR DESCRIPTION
**Pull request by LeeYang2019**
- Fixes #104: added query search feature.

## Changes
- Change 1: created search query to return a list of users based on firstname, lastname or email
- Change 2: wrote tests to return singular and multiple individuals based on firstname and lastname

## PR Checklist
- [x] Does your PR contain relevant tests that pass?
- [x] Is your code linted? `yarn lint` or `yarn lint:fix`
- [x] Does your code build correctly in ALL projects? `yarn build`

### CODE REVIEW

File: packages/api/src/User/User.resolver.ts [hunk 1]
```diff
@@ -1,27 +1,42 @@
...
 
   @UseGuards(GQLAuthGuard)
   @Query(() => [User])
   users() {
     return this.userService.findAll();
   }
 
+  @UseGuards(GQLAuthGuard)
+  @Query(() => [User])
+  search(@Args('param') param: string) {
```
Comment (quimgv) [inline packages/api/src/User/User.resolver.ts:36]: Let's name this searchUsers

File: packages/api/tests/utils/queries/search.gql [hunk 2]
```diff
@@ -0,0 +1,17 @@
+query search($param: String!) {
```
Comment (quimgv) [inline packages/api/tests/utils/queries/search.gql:1]: same here: searchUsers

File: packages/api/tests/utils/queries/index.ts [hunk 3]
```diff
@@ -1,8 +1,10 @@
...
 import path from 'path';
 
 export default {
-  getPathByName: fs.readFileSync(path.join(__dirname, './getPathByName.gql')).toString(),
-  me: fs.readFileSync(path.join(__dirname, './me.gql')).toString()
+  getPathByName: fs
+    .readFileSync(path.join(__dirname, './getPathByName.gql'))
+    .toString(),
+  me: fs.readFileSync(path.join(__dirname, './me.gql')).toString(),
+  search: fs.readFileSync(path.join(__dirname, './search.gql')).toString(),
```
Comment (quimgv) [inline packages/api/tests/utils/queries/index.ts:9]: and here (searchUsers)

File: packages/api/src/User/User.resolver.ts (hunk 1 above)
Comment (LeeYang2019) [reply packages/api/src/User/User.resolver.ts:36]: This has been reworded to searchUsers.

File: packages/api/tests/utils/queries/search.gql (hunk 2 above)
Comment (LeeYang2019) [reply packages/api/tests/utils/queries/search.gql:1]: This has been reworded to searchUsers.

File: packages/api/tests/utils/queries/index.ts (hunk 3 above)
Comment (LeeYang2019) [reply packages/api/tests/utils/queries/index.ts:9]: This has been reworded to searchUsers.

File: packages/api/tests/utils/queries/search.gql (hunk 2 above)
Comment (quimgv) [reply packages/api/tests/utils/queries/search.gql:1]: You need to push again @LeeYang2019

**Comment by quimgv** [comment]
@tristanMatthias isn't this ready to merge?

**Comment by shaneSheridan** [comment]
@quimgv It says that you've requested changes, that says to me that the PR owner needs to change code.

Or are you just requesting @tristanMatthias to review?

**Comment by quimgv** [comment]
@shaneSheridan Yes, I've sent you and update through Slack :)

File: packages/api/src/User/User.resolver.ts [hunk 4]
```diff
@@ -1,27 +1,42 @@
...
 
   @UseGuards(GQLAuthGuard)
   @Query(() => [User])
   users() {
     return this.userService.findAll();
   }
 
+  @UseGuards(GQLAuthGuard)
+  @Query(() => [User])
+  searchUsers(@Args('param') param: string) {
```
Comment (tristanMatthias) [inline packages/api/src/User/User.resolver.ts:36]: Can you rename this to `query`? `param` is too generic

File: packages/api/src/User/User.resolver.ts [hunk 5]
```diff
@@ -48,7 +64,9 @@ export class UserResolver {
 
   @ResolveField(() => UserPreferences)
   async userPreferences(@Parent() user: User) {
-    const userPreferences = await this.userPreferencesService.findByUser(user.id);
+    const userPreferences = await this.userPreferencesService.findByUser(
+      user.id
+    );
     return userPreferences;
```
Comment (tristanMatthias) [inline packages/api/src/User/User.resolver.ts:70]: Small thing, but you can just combine these into one line. EG: `return await this.UserPref....`

File: packages/api/src/User/User.resolver.ts (hunk 4 above)
Comment (LeeYang2019) [reply packages/api/src/User/User.resolver.ts]: Hey @tristanMatthias, "param" has been updated to "query".

File: packages/api/src/User/User.resolver.ts (hunk 5 above)
Comment (LeeYang2019) [reply packages/api/src/User/User.resolver.ts]: Hey @tristanMatthias, code has been refactored.

File: packages/api/src/User/User.resolver.ts (hunk 4 above)
Comment (tristanMatthias) [reply packages/api/src/User/User.resolver.ts]: Did you push? I can't see this yet

**Comment by tristanMatthias** [comment]
Hey mate, looking good. There's been a few things changed in develop, so if you want to merge that in and fix the conflicts, then we should be good to go!

File: packages/api/src/User/User.resolver.ts (hunk 4 above)
Comment (shaneSheridan) [reply packages/api/src/User/User.resolver.ts]: @LeeYang2019 Did you see the above comment from @tristanMatthias?
Looks like you haven't pushed your updated code yet.

**Comment by shaneSheridan** [comment]
@LeeYang2019 Looks like some code changes are still needed as per the review comments, also the conflicting files need to be resolved.
Let me know if you need to support with this.

**Comment by quimgv** [comment]
@shaneSheridan @tristanMatthias I merged develop and fixed the conflicts which all of them were tests and they are not passing because Lee and some others are currently working on them in another branch. So this can be merged and they will fix the tests in the branch they are currently working on now.

**Comment by tristanMatthias** [comment]
Once these tests pass lets merge in. If you're doing these in another branch, let me know and I'll merge this in now

**Comment by LeeYang2019** [comment]
Hey @tristanMatthias, can we merge this? Is there anything else holding this up? Thanks.

**Comment by tristanMatthias** [comment]
Yeah the merge conflicts and tests are failing on this. Are they passing locally?

**Comment by quimgv** [comment]
Ready to merge @tristanMatthias

**Pull request merged**